### PR TITLE
fix: replace `react-window` with `react-virtuoso` in transfer component

### PR DESCRIPTION
**Pull request by keita-determined**
## Description
[WEB-1958]

## Test Plan
- Check if transfer component is virtualized and work as expected in `binding tab in resource pool (RBAC required)` and `column button in the old experiment table`
  - should be able to add/remove items by clicking items or `add/remove all` in the bottom
  - should be able to search items by text and text should be highlighted
  - should be able to drag & drop items to reorder columns in `visible` form

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[WEB-1958]: https://hpe-aiatscale.atlassian.net/browse/WEB-1958?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/components/Transfer.module.scss [hunk 1]
```diff
@@ -29,13 +29,15 @@
         ul {
           height: 200px;
           list-style-type: none;
+          margin: 0;
           margin-bottom: 0;
```
Comment (EmilyBonar) [inline webui/react/src/components/Transfer.module.scss:33]: nit: if you're setting `margin: 0` you can remove `margin-bottom: 0`

File: webui/react/src/components/Transfer.tsx [hunk 2]
```diff
@@ -189,15 +182,16 @@ const Transfer: React.FC<Props> = ({
...
-            className={css.listContainer}
-            height={200}
-            innerElementType="ul"
-            itemCount={filteredHiddenEntries.length}
-            itemSize={24}
-            width="100%">
-            {renderHiddenRow}
-          </List>
+          <div className={css.listContainer} ref={rightRef}>
+            <ul>
```
Comment (EmilyBonar) [inline webui/react/src/components/Transfer.tsx:186]: nit: you can combine these two lines into one `ul` with the className and ref

File: webui/react/src/components/Transfer.tsx [hunk 3]
```diff
@@ -207,15 +201,16 @@ const Transfer: React.FC<Props> = ({
...
-            className={css.listContainer}
-            height={200}
-            innerElementType="ul"
-            itemCount={filteredVisibleEntries.length}
-            itemSize={24}
-            width="100%">
-            {renderVisibleRow}
-          </List>
+          <div className={css.listContainer} ref={leftRef}>
+            <ul>
```
Comment (EmilyBonar) [inline webui/react/src/components/Transfer.tsx:205]: nit: same note on combining

File: webui/react/src/components/Transfer.tsx [hunk 4]
```diff
@@ -189,15 +182,16 @@ const Transfer: React.FC<Props> = ({
...
-            innerElementType="ul"
-            itemCount={filteredHiddenEntries.length}
-            itemSize={24}
-            width="100%">
-            {renderHiddenRow}
-          </List>
+          <div className={css.listContainer} ref={rightRef}>
+            <ul>
+              <Virtuoso
+                customScrollParent={rightRef.current ?? undefined}
```
Comment (EmilyBonar) [inline webui/react/src/components/Transfer.tsx:188]: q: do you need a `customScrollParent` here? if not, remove from the other list as well

File: webui/react/src/components/Transfer.tsx [hunk 5]
```diff
@@ -166,19 +159,19 @@ const Transfer: React.FC<Props> = ({
   );
 
   const renderHiddenRow = useCallback(
-    ({ index, style }: { index: number; style: React.CSSProperties }) => {
+    ({ index }: { index: number }) => {
```
Comment (EmilyBonar) [inline webui/react/src/components/Transfer.tsx:162]: nit: you can simplify this line (and the one in `renderVisibleRow`) to just `(index: number) => {`

File: webui/react/src/components/Transfer.tsx [hunk 6]
```diff
@@ -116,48 +119,38 @@ const Transfer: React.FC<Props> = ({
...
-    },
-    [targetEntries],
-  );
+        const temp = updatedVisibleEntries[entryIndex];
+        updatedVisibleEntries[entryIndex] = updatedVisibleEntries[newNeighborEntryIndex];
+        updatedVisibleEntries[newNeighborEntryIndex] = temp;
+        return updatedVisibleEntries;
+      });
+    }
+    return;
```
Comment (EmilyBonar) [inline webui/react/src/components/Transfer.tsx:142]: q: is this return necessary?

File: webui/react/src/components/ColumnsCustomizeModal.test.tsx [hunk 7]
```diff
@@ -222,15 +229,11 @@ describe('Columns Customize Modal', () => {
...
-
     await user.click(await view.findByText('Add All'));
 
     await waitFor(() => {
-      expect(parseInt(lists[0].style.height)).toEqual(0);
-      expect(parseInt(lists[1].style.height)).toEqual(
-        (NUM_GENERATED_COLUMNS + DEFAULT_COLUMNS.length) * lineHeight,
-      );
+      expect(lists[0].children[0].children[0].children[0].childElementCount).toEqual(0);
+      expect(lists[1].children[0].children[0].children[0].childElementCount).toBeGreaterThan(0);
```
Comment (EmilyBonar) [inline webui/react/src/components/ColumnsCustomizeModal.test.tsx:236]: nit: these are some nasty selectors, can you look into some way of testing this that doesn't require drilling so deep into child elements?

**Review by EmilyBonar — CHANGES_REQUESTED** [review]
The code mostly looks good, though a few changes could clean it up more. Can you provide a better test plan so I can actually try it out?

File: webui/react/src/components/Transfer.tsx (hunk 4 above)
Comment (keita-determined) [reply webui/react/src/components/Transfer.tsx]: didnt need it, so removed it

File: webui/react/src/components/Transfer.tsx (hunk 6 above)
Comment (keita-determined) [reply webui/react/src/components/Transfer.tsx]: leftover from the old code. removed it

File: webui/react/src/components/ColumnsCustomizeModal.test.tsx (hunk 7 above)
Comment (keita-determined) [reply webui/react/src/components/ColumnsCustomizeModal.test.tsx]: used `querySelector` to refactor this code

**Review by EmilyBonar — CHANGES_REQUESTED** [review]
Something is broken in the drag and drop handling, I ended up in this state when I tried to rearrange columns.
Screenshot 2024-02-13 at 11-59-18 Uncategorized Experiments - Determined

**Comment by EmilyBonar** [comment]
Also the capitalization of the column names should be standardized.

**Comment by keita-determined** [comment]
@EmilyBonar do you have the replicate steps? I couldnt reproduce it on chrome and do you see the issue on main branch?

**Comment by EmilyBonar** [comment]
No specific steps to replicate, just dragging and dropping columns to rearrange. Can't repro on latest-main. I'm seeing the bug on Firefox.

**Comment by keita-determined** [comment]
Surprisingly, i cannot reproduce it on Firefox either. I see netlify is not working well on any PRs. Probably that could be the issue??

**Comment by keita-determined** [comment]
also texts are not capitalized since these texts are used as id to track the columns. I think we should resolve it since it is not related to this ticket.

**Comment by keita-determined** [comment]
@EmilyBonar i think i could fix the issue. could you review the pr? thanks

**Review by EmilyBonar — APPROVED** [review]
LGTM!

**Pull request merged**